httpapi/httpapi.go

```diff
@@ -125,6 +142,64 @@ func recordListHandler(hy *hydra.Hydra) func(http.ResponseWriter, *http.Request)
 	}
 }
 
+// "/records/add" Add new records
+func recordAddHandler(hy *hydra.Hydra) func(http.ResponseWriter, *http.Request) {
+	return func(w http.ResponseWriter, r *http.Request) {
+		// Verify the Content-Type matches
+		contentTypes := r.Header["Content-Type"]
+
+		if len(contentTypes) < 1 || !strings.HasPrefix(contentTypes[0], "application/json") {
+			writeApiErrorResponse(
+				w, http.StatusUnsupportedMediaType, "Request must specify the Content-Type header as \"application/json\".",
+			)
+			return
+		}
+
+		// Decode the body as JSON
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			fmt.Printf("Error on adding provider records while reading the request payload: %s\n", err)
+			writeApiErrorResponse(w, http.StatusInternalServerError, "")
+			return
+		}
+
+		var rawRecords []rawProviderRecord
+		err = json.Unmarshal(body, &rawRecords)
+
+		if err != nil {
+			writeApiErrorResponse(w, http.StatusBadRequest, "Invalid request payload.")
+			return
+		}
+
+		// Prepare the records to add - We don't add directly as we want to operation to fail if any of the record is invalid
+		records := make([]providerRecord, len(rawRecords))
+		for i, rawRecord := range rawRecords {
+			cid, err := cid.Decode(rawRecord.CID)
+
+			if err != nil {
+				writeApiErrorResponse(w, http.StatusBadRequest, fmt.Sprintf("Invalid CID provided on record[%d].", i))
+				return
+			}
+
+			peerId, err := peer.Decode(rawRecord.PeerID)
+			if err != nil {
+				writeApiErrorResponse(w, http.StatusBadRequest, fmt.Sprintf("Invalid PeerID provided on record[%d].", i))
+				return
+			}
+
+			records[i] = providerRecord{cid, peerId}
+		}
+
+		// Store the new records locally
+		ctx := r.Context()
+		for _, record := range records {
+			hy.Heads[0].AddProvider(ctx, record.CID, record.PeerID)
```

Comment:
This is correct, but I noticed that the AddProvider method is not, it converts the CID to the raw CID bytes to use as the DHT key, however DHT clients expect DHT keys to be multihashes (see e.g. https://github.com/libp2p/go-libp2p-kad-dht/blob/master/routing.go#L477), which means this won't work for CIDv1.  This code itself is fine though. I'll open a separate PR to fix AddProvider.